v2/apiserver/internal/core/events.go

```diff
@@ -637,33 +642,38 @@ func (e *eventsService) DeleteMany(
 		)
 	}
 
-	events, err := e.eventsStore.DeleteMany(ctx, selector)
+	eventCh, affectedCount, err := e.eventsStore.DeleteMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error deleting events from store")
 	}
-
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker way? Or
-	// if not quicker, more reliable?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
+	result.Count = affectedCount
+
+	// Fan out to a finite number of goroutines to handle cleanup duties
+	concurrency := 10
+	if result.Count < int64(concurrency) {
+		concurrency = int(result.Count)
+	}
+	for i := 0; i < concurrency; i++ {
+		go func() {
+			for {
+				event, ok := <-eventCh
+				if !ok { // event channel closed; we're done
+					return
+				}
```

Comment:
Nit: These five lines can be condensed to this. (I wasn't sure at first, but I tested it. You never get an event that is a zero value `Event{}`.)

```
for event := range  eventCh {
```

Could be applied in one other spot as well.

Reply:
_But_ this is totally non-blocking. LGTM!

Reply:
Oooh nice, thanks for testing and proposing.  Implemented.